### ISSUE TITLE
Build fail: error: inlining failed in call to 'always_inline' 'openat': function body can be overwritten at link time

### ISSUE THREAD

**Issue by petterreinholdtsen** [davidel/fl-cow#1]
When trying to build version 0.10 on Debian Bookworm and Sid, the compiler reject the code like this:

```
libtool: compile:  gcc -DHAVE_CONFIG_H -I. -I.. -I. -I.. -Wdate-time -D_FORTIFY_SOURCE=2 -Wstrict-prototypes -pipe -g -O2 -Werror=implicit-function-declaration -ffile-prefix-map=/home/pere/src/debian/fl-cow-salsa=. -fstack-protector-strong -fstack-clash-protection -Wformat -Werror=format-security -fcf-protection -c fl-cow.c  -fPIC -DPIC -o .libs/libflcow_la-fl-cow.o
fl-cow.c: In function 'do_cow_name':
fl-cow.c:205:9: warning: ignoring return value of 'fchown' declared with attribute 'warn_unused_result' [-Wunused-result]
  205 |         fchown(nfd, stb.st_uid, stb.st_gid);
      |         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /usr/include/fcntl.h:341,
                 from fl-cow.c:28:
/usr/include/x86_64-linux-gnu/bits/fcntl2.h:159:1: error: inlining failed in call to 'always_inline' 'openat': function body can be overwritten at link time
  159 | openat (int __fd, const char *__path, int __oflag, ...)
      | ^~~~~~
fl-cow.c:178:20: note: called from here
  178 |         if ((sfd = openat(dirfd, name, O_RDONLY, 0)) == -1)
      |                    ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/include/x86_64-linux-gnu/bits/fcntl2.h:43:1: error: inlining failed in call to 'always_inline' 'open': function body can be overwritten at link time
   43 | open (const char *__path, int __oflag, ...)
      | ^~~~
fl-cow.c:185:20: note: called from here
  185 |         if ((nfd = open(fpath, O_CREAT | O_EXCL | O_WRONLY, stb.st_mode)) == -1) {
      |                    ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

Any idea how to get the code to build?

**Comment by davidel** [comment]
I have modernized autoconf machinery in the latest commit but this was building even with the previous one.
Can you provide the steps you used to build, and also exact Debian and GCC versions?

In my machine (Debian 12.8 and GCC 12.2.0:

```
davide@davide-lnx1:/tmp$ git clone https://github.com/davidel/fl-cow.git
Cloning into 'fl-cow'...
remote: Enumerating objects: 71, done.
remote: Counting objects: 100% (71/71), done.
remote: Compressing objects: 100% (49/49), done.
remote: Total 71 (delta 22), reused 60 (delta 16), pack-reused 0 (from 0)
Receiving objects: 100% (71/71), 486.60 KiB | 1.73 MiB/s, done.
Resolving deltas: 100% (22/22), done.
```
```
davide@davide-lnx1:/tmp$ cd fl-cow/
```
```
davide@davide-lnx1:/tmp/fl-cow$ ./configure 
checking for a BSD-compatible install... /usr/bin/install -c
checking whether build environment is sane... yes
checking for a race-free mkdir -p... /usr/bin/mkdir -p
checking for gawk... no
checking for mawk... mawk
checking whether make sets $(MAKE)... yes
checking whether make supports nested variables... yes
checking whether make sets $(MAKE)... (cached) yes
checking whether make supports the include directive... yes (GNU style)
checking for gcc... gcc
checking whether the C compiler works... yes
checking for C compiler default output file name... a.out
checking for suffix of executables... 
checking whether we are cross compiling... no
checking for suffix of object files... o
checking whether the compiler supports GNU C... yes
checking whether gcc accepts -g... yes
checking for gcc option to enable C11 features... none needed
checking whether gcc understands -c and -o together... yes
checking dependency style of gcc... gcc3
checking for stdio.h... yes
checking for stdlib.h... yes
checking for string.h... yes
checking for inttypes.h... yes
checking for stdint.h... yes
checking for strings.h... yes
checking for sys/stat.h... yes
checking for sys/types.h... yes
checking for unistd.h... yes
checking for wchar.h... yes
checking for minix/config.h... no
checking whether it is safe to define __EXTENSIONS__... yes
checking whether _XOPEN_SOURCE should be defined... no
checking for gcc... (cached) gcc
checking whether the compiler supports GNU C... (cached) yes
checking whether gcc accepts -g... (cached) yes
checking for gcc option to enable C11 features... (cached) none needed
checking whether gcc understands -c and -o together... (cached) yes
checking dependency style of gcc... (cached) gcc3
checking build system type... x86_64-pc-linux-gnu
checking host system type... x86_64-pc-linux-gnu
checking how to print strings... printf
checking for a sed that does not truncate output... /usr/bin/sed
checking for grep that handles long lines and -e... /usr/bin/grep
checking for egrep... /usr/bin/grep -E
checking for fgrep... /usr/bin/grep -F
checking for ld used by gcc... /usr/bin/ld
checking if the linker (/usr/bin/ld) is GNU ld... yes
checking for BSD- or MS-compatible name lister (nm)... /usr/bin/nm -B
checking the name lister (/usr/bin/nm -B) interface... BSD nm
checking whether ln -s works... yes
checking the maximum length of command line arguments... 1572864
checking how to convert x86_64-pc-linux-gnu file names to x86_64-pc-linux-gnu format... func_convert_file_noop
checking how to convert x86_64-pc-linux-gnu file names to toolchain format... func_convert_file_noop
checking for /usr/bin/ld option to reload object files... -r
checking for file... file
checking for objdump... objdump
checking how to recognize dependent libraries... pass_all
checking for dlltool... no
checking how to associate runtime and link libraries... printf %s\n
checking for ar... ar
checking for archiver @FILE support... @
checking for strip... strip
checking for ranlib... ranlib
checking command to parse /usr/bin/nm -B output from gcc object... ok
checking for sysroot... no
checking for a working dd... /usr/bin/dd
checking how to truncate binary pipes... /usr/bin/dd bs=4096 count=1
checking for mt... mt
checking if mt is a manifest tool... no
checking for dlfcn.h... yes
checking for objdir... .libs
checking if gcc supports -fno-rtti -fno-exceptions... no
checking for gcc option to produce PIC... -fPIC -DPIC
checking if gcc PIC flag -fPIC -DPIC works... yes
checking if gcc static flag -static works... yes
checking if gcc supports -c -o file.o... yes
checking if gcc supports -c -o file.o... (cached) yes
checking whether the gcc linker (/usr/bin/ld -m elf_x86_64) supports shared libraries... yes
checking whether -lc should be explicitly linked in... no
checking dynamic linker characteristics... GNU/Linux ld.so
checking how to hardcode library paths into programs... immediate
checking whether stripping libraries is possible... yes
checking if libtool supports shared libraries... yes
checking whether to build shared libraries... yes
checking whether to build static libraries... yes
checking for library containing dlopen... none required
checking for egrep... (cached) /usr/bin/grep -E
checking for inline... inline
checking for working volatile... yes
checking for an ANSI C-conforming const... yes
checking whether byte ordering is bigendian... no
checking for dladdr... yes
checking for open64... yes
checking for memset... yes
checking for memcpy... yes
checking for getcwd... yes
checking for strlen... yes
checking for strncmp... yes
checking that generated files are newer than configure... done
configure: creating ./config.status
config.status: creating Makefile
config.status: creating fl-cow/Makefile
config.status: creating test/Makefile
config.status: creating config.h
config.status: executing depfiles commands
config.status: executing libtool commands
```
```
davide@davide-lnx1:/tmp/fl-cow$ make
CDPATH="${ZSH_VERSION+.}:" && cd . && /bin/bash '/tmp/fl-cow/missing' aclocal-1.16 -I m4
aclocal-1.16: warning: couldn't open directory 'm4': No such file or directory
 cd . && /bin/bash /tmp/fl-cow/missing automake-1.16 --gnu
CDPATH="${ZSH_VERSION+.}:" && cd . && /bin/bash '/tmp/fl-cow/missing' autoconf
/bin/bash ./config.status --recheck
running CONFIG_SHELL=/bin/bash /bin/bash ./configure --no-create --no-recursion
checking for a BSD-compatible install... /usr/bin/install -c
checking whether build environment is sane... yes
checking for a race-free mkdir -p... /usr/bin/mkdir -p
checking for gawk... no
checking for mawk... mawk
checking whether make sets $(MAKE)... yes
checking whether make supports nested variables... yes
checking whether make sets $(MAKE)... (cached) yes
checking whether make supports the include directive... yes (GNU style)
checking for gcc... gcc
checking whether the C compiler works... yes
checking for C compiler default output file name... a.out
checking for suffix of executables... 
checking whether we are cross compiling... no
checking for suffix of object files... o
checking whether the compiler supports GNU C... yes
checking whether gcc accepts -g... yes
checking for gcc option to enable C11 features... none needed
checking whether gcc understands -c and -o together... yes
checking dependency style of gcc... gcc3
checking for stdio.h... yes
checking for stdlib.h... yes
checking for string.h... yes
checking for inttypes.h... yes
checking for stdint.h... yes
checking for strings.h... yes
checking for sys/stat.h... yes
checking for sys/types.h... yes
checking for unistd.h... yes
checking for wchar.h... yes
checking for minix/config.h... no
checking whether it is safe to define __EXTENSIONS__... yes
checking whether _XOPEN_SOURCE should be defined... no
checking for gcc... (cached) gcc
checking whether the compiler supports GNU C... (cached) yes
checking whether gcc accepts -g... (cached) yes
checking for gcc option to enable C11 features... (cached) none needed
checking whether gcc understands -c and -o together... (cached) yes
checking dependency style of gcc... (cached) gcc3
checking build system type... x86_64-pc-linux-gnu
checking host system type... x86_64-pc-linux-gnu
checking how to print strings... printf
checking for a sed that does not truncate output... /usr/bin/sed
checking for grep that handles long lines and -e... /usr/bin/grep
checking for egrep... /usr/bin/grep -E
checking for fgrep... /usr/bin/grep -F
checking for ld used by gcc... /usr/bin/ld
checking if the linker (/usr/bin/ld) is GNU ld... yes
checking for BSD- or MS-compatible name lister (nm)... /usr/bin/nm -B
checking the name lister (/usr/bin/nm -B) interface... BSD nm
checking whether ln -s works... yes
checking the maximum length of command line arguments... 1572864
checking how to convert x86_64-pc-linux-gnu file names to x86_64-pc-linux-gnu format... func_convert_file_noop
checking how to convert x86_64-pc-linux-gnu file names to toolchain format... func_convert_file_noop
checking for /usr/bin/ld option to reload object files... -r
checking for file... file
checking for objdump... objdump
checking how to recognize dependent libraries... pass_all
checking for dlltool... no
checking how to associate runtime and link libraries... printf %s\n
checking for ar... ar
checking for archiver @FILE support... @
checking for strip... strip
checking for ranlib... ranlib
checking command to parse /usr/bin/nm -B output from gcc object... ok
checking for sysroot... no
checking for a working dd... /usr/bin/dd
checking how to truncate binary pipes... /usr/bin/dd bs=4096 count=1
checking for mt... mt
checking if mt is a manifest tool... no
checking for dlfcn.h... yes
checking for objdir... .libs
checking if gcc supports -fno-rtti -fno-exceptions... no
checking for gcc option to produce PIC... -fPIC -DPIC
checking if gcc PIC flag -fPIC -DPIC works... yes
checking if gcc static flag -static works... yes
checking if gcc supports -c -o file.o... yes
checking if gcc supports -c -o file.o... (cached) yes
checking whether the gcc linker (/usr/bin/ld -m elf_x86_64) supports shared libraries... yes
checking whether -lc should be explicitly linked in... no
checking dynamic linker characteristics... GNU/Linux ld.so
checking how to hardcode library paths into programs... immediate
checking whether stripping libraries is possible... yes
checking if libtool supports shared libraries... yes
checking whether to build shared libraries... yes
checking whether to build static libraries... yes
checking for library containing dlopen... none required
checking for egrep... (cached) /usr/bin/grep -E
checking for inline... inline
checking for working volatile... yes
checking for an ANSI C-conforming const... yes
checking whether byte ordering is bigendian... no
checking for dladdr... yes
checking for open64... yes
checking for memset... yes
checking for memcpy... yes
checking for getcwd... yes
checking for strlen... yes
checking for strncmp... yes
checking that generated files are newer than configure... done
configure: creating ./config.status
 /bin/bash ./config.status
config.status: creating Makefile
config.status: creating fl-cow/Makefile
config.status: creating test/Makefile
config.status: creating config.h
config.status: config.h is unchanged
config.status: executing depfiles commands
config.status: executing libtool commands
(CDPATH="${ZSH_VERSION+.}:" && cd . && /bin/bash '/tmp/fl-cow/missing' autoheader)
rm -f stamp-h1
touch config.h.in
cd . && /bin/bash ./config.status config.h
config.status: creating config.h
config.status: config.h is unchanged
make  all-recursive
make[1]: Entering directory '/tmp/fl-cow'
Making all in .
make[2]: Entering directory '/tmp/fl-cow'
make[2]: Leaving directory '/tmp/fl-cow'
Making all in fl-cow
make[2]: Entering directory '/tmp/fl-cow/fl-cow'
/bin/bash ../libtool  --tag=CC   --mode=compile gcc -DHAVE_CONFIG_H -I. -I..    -I, -I.. -Wstrict-prototypes -pipe -g -O2 -MT libflcow_la-fl-cow.lo -MD -MP -MF .deps/libflcow_la-fl-cow.Tpo -c -o libflcow_la-fl-cow.lo `test -f 'fl-cow.c' || echo './'`fl-cow.c
libtool: compile:  gcc -DHAVE_CONFIG_H -I. -I.. -I, -I.. -Wstrict-prototypes -pipe -g -O2 -MT libflcow_la-fl-cow.lo -MD -MP -MF .deps/libflcow_la-fl-cow.Tpo -c fl-cow.c  -fPIC -DPIC -o .libs/libflcow_la-fl-cow.o
libtool: compile:  gcc -DHAVE_CONFIG_H -I. -I.. -I, -I.. -Wstrict-prototypes -pipe -g -O2 -MT libflcow_la-fl-cow.lo -MD -MP -MF .deps/libflcow_la-fl-cow.Tpo -c fl-cow.c -o libflcow_la-fl-cow.o >/dev/null 2>&1
mv -f .deps/libflcow_la-fl-cow.Tpo .deps/libflcow_la-fl-cow.Plo
/bin/bash ../libtool  --tag=CC   --mode=link gcc -I, -I.. -Wstrict-prototypes -pipe -g -O2 -module -avoid-version  -o libflcow.la -rpath /usr/local/lib libflcow_la-fl-cow.lo  
libtool: link: gcc -shared  -fPIC -DPIC  .libs/libflcow_la-fl-cow.o    -g -O2   -Wl,-soname -Wl,libflcow.so -o .libs/libflcow.so
libtool: link: ar cr .libs/libflcow.a  libflcow_la-fl-cow.o
libtool: link: ranlib .libs/libflcow.a
libtool: link: ( cd ".libs" && rm -f "libflcow.la" && ln -s "../libflcow.la" "libflcow.la" )
make[2]: Leaving directory '/tmp/fl-cow/fl-cow'
Making all in test
make[2]: Entering directory '/tmp/fl-cow/test'
gcc -DHAVE_CONFIG_H -I. -I..     -g -O2 -MT flcow-test.o -MD -MP -MF .deps/flcow-test.Tpo -c -o flcow-test.o flcow-test.c
mv -f .deps/flcow-test.Tpo .deps/flcow-test.Po
/bin/bash ../libtool  --tag=CC   --mode=link gcc  -g -O2   -o flcow-test flcow-test.o  
libtool: link: gcc -g -O2 -o flcow-test flcow-test.o 
make[2]: Leaving directory '/tmp/fl-cow/test'
make[1]: Leaving directory '/tmp/fl-cow'
```
```
davide@davide-lnx1:/tmp/fl-cow$ cat /etc/debian_version 
12.8
```
```
davide@davide-lnx1:/tmp/fl-cow$ gcc --version
gcc (Debian 12.2.0-14) 12.2.0
Copyright (C) 2022 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
```

**Comment by petterreinholdtsen** [comment]
[Davide Libenzi]
> Can you provide the steps you used to build, and also exact Debian and
> GCC versions?

Here is a recipe to reproduce the failure I see on Debian Bookworm and
Sid, reproduced from memory, might have typos:

  apt install devscripts git-buildpackage
  gbp clone https://salsa.debian.org/debian/fl-cow.git
  uscan -v
  gbp import-orig fl-cow_0.10.orig.tar.gz
  vi debian/patches/series # comment out all patches, no longer apply
  gbp dch --qa
  apt build-dep .
  debuild

This will fetch the current Debian built rules, fetch the latest
upstream tarball, import it into the git repository, disable patches no
longer applying cleanly (need to be investigated to see if they are
still relevant), update debian/changelog with new version info, install
build dependencies and try to build the source.

% gcc --version
gcc (Debian 12.2.0-14) 12.2.0
Copyright (C) 2022 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

% lsb_release -a
No LSB modules are available.
Distributor ID: Debian
Description:    Debian GNU/Linux 12 (bookworm)
Release:        12
Codename:       bookworm
% cat /etc/debian_version 
12.8
%

-- 
Happy hacking
Petter Reinholdtsen

**Comment by davidel** [comment]
Hi!
I have provided the zero-dependency (besides git and make) steps that shows the the library builds correctly.
I believe it is not the task of a library maintainer to understand what every distro build chain does (I gave a try to the steps above - which had errors themselves - with no luck, given they complained about a mismatch version of 0.6 vs 0.10), and debug that.

The suggestion I give you, is to drop this package from Debian, since it is a very old software.
I used that (together with hard-link copies and quilt - an old patch managing tool) to work on the linux kernel before GIT came into existence.
Today there are much better tools (overlayfs and fuse-based implementations come first to my mind) to do what this library does, and safer too (GLIBC might have added other pathways to file opens, which might escape the fl-cow hooks).

**Comment by petterreinholdtsen** [comment]
[Davide Libenzi]
> given they complained about a mismatch version of 0.6 vs 0.10), and
> debug that.

Sound like the 'gbp dch --qa' step did not insert the correct version
number in debian/changelog.

> The suggestion I give you, is to drop this package from Debian, since
> it is a very old software.

No worries.  My current project is to give orphaned Debian packages
missing a working git repository for the packaging a git repository on
salsa.debian.org.  As part of this process I try to leave the package a
bit better than when I found it, and try to let upstream know about any
patches lingering in Debian.  This is where you came in. :)

As you responded quickly and mentioned a new version, I decided to spend
a bit extra time on the fl-cow package to respond in kind to your kind
feedback, but do not plan to start deciding which packages to remove.

-- 
Happy hacking
Petter Reinholdtsen
